model/incoming_webhook.go

```diff
@@ -63,11 +98,201 @@ func IncomingWebhookListFromJson(data io.Reader) []*IncomingWebhook {
 	return o
 }
 
+// TN added, this should be modified
+
+func (o *IncomingWebhook) ResetModels() {
+	o.HmacAlgorithm	= ""
+	o.TimestampModel = StringInterface{}
+	o.HmacModel = StringInterface{}
+	o.SignedContentModel = StringArray{}
+}
+
+func ValidModelSplitter(s *StringInterface) bool {
+	// empty splitBy is valid
+	splitValue, isSet := s.ValidHeaderModel("SplitBy")
+	if ! isSet {
+		delete(*s, "Index")
+		return true
+	}
+
+	idxValue,_ := s.ValidHeaderModel("Index")
+	if len(splitValue) > 0{
+		if ! (len(idxValue) > 0) {
+			return false
+		}
+		intIdx,err:= strconv.ParseInt(idxValue, 10, 0)
+		if err != nil {
+			return false
+		} else if int(intIdx) > 5 || int(intIdx) < 0 {
+			return false
+		}
+	} else {
+		if len(idxValue) > 0 {
+			return false
+		}
+	}
+
+	return true
+}
+
+func (s StringInterface) ValidHeaderModel(mapKey string) (string, bool) {
+	var isSet bool = false
+	var val string = ""
+	if s[mapKey] != nil {
+		isSet = true
+	}
+	if headerPart,ok := s[mapKey].(string); ok {
+		val = headerPart
+	}
+	return val, isSet
+} 
+
+func (o *IncomingWebhook) IsValidModels() bool {
+	if ! (o.HmacAlgorithm == "HMAC-SHA1" || o.HmacAlgorithm == "HMAC-SHA256") {
+		return false
+	}
+
+	if ! (len(o.SignedContentModel) > 0 && len(o.SignedContentModel) < 6) {
+		return false
+	}
+
+	hmacModelHeader, ok := o.HmacModel["HeaderName"].(string)
+	if  ! (ok && 0 < len(hmacModelHeader) && len(hmacModelHeader) < 26) {
+		return false
+	}
+	if ! ValidModelSplitter(&o.HmacModel) {
+		return false
+	}
+
+	if pre,pSet := o.HmacModel.ValidHeaderModel("Prefix"); pSet && pre == "" {
+		return false
+	}
+
+	timestampModelH := o.TimestampModel["HeaderName"]
+	if timestampModelH == nil {
+		o.TimestampModel = StringInterface{}
+	} else {
+		tsModelHeader, ok1 := timestampModelH.(string)
+		if  ! (ok1 && 0 < len(tsModelHeader) && len(tsModelHeader) < 26) {
+			return false
+
+		}
+
+		if ! ValidModelSplitter(&o.TimestampModel) {
+			return false
+		}
+
+		if pre,pSet := o.TimestampModel.ValidHeaderModel("Prefix"); pSet && (pre == "" || len(pre) > 8) {
+			return false
+		}
+		
+	}
+
+	return true
+}
+
+// this should be replaced with a better way
+func ParseHeaderModel(signingPart StringInterface) (HeaderModel, *AppError) {
```

Comment:
I do not understand what this function does. Could you explain it a little further?

Reply:
@larkox, this was one of the todo list in the description above. It parses StringInterface to HeaderModel struct by using json. I need a little function to parse it so that there is no ugly json marshal and unmarshal. This is expected in next commit. Thanks again